File: packages/gavio-js/src/shim/openai.ts
/**
 * OpenAI drop-in shim (F-DX-04) — point existing OpenAI SDK code at Gavio.
 *
 *   import { Gateway } from 'gavio'
 *   import { GavioOpenAI } from 'gavio/shim/openai'
 *
 *   const client = new GavioOpenAI(new Gateway({ provider: 'openai', model: 'gpt-4o' }))
 *   const resp = await client.chat.completions.create({
 *     model: 'gpt-4o', messages: [{ role: 'user', content: 'hi' }],
 *   })
 *   console.log(resp.choices[0].message.content)
 */

import type { Gateway } from '../gateway.js'
import type { GavioResponse } from '../response.js'
import type { Message } from '../types.js'

export interface ChatCompletion {
  id: string
  object: 'chat.completion'
  model: string
  choices: { index: number; message: { role: string; content: string }; finish_reason: string }[]
  usage: { prompt_tokens: number; completion_tokens: number; total_tokens: number }
  gavio: { costUsd: number; cacheHit: boolean; interceptorsFired: string[] }
}

export interface CreateParams {
  messages: Message[]
  model?: string
  temperature?: number
  // eslint-disable-next-line @typescript-eslint/naming-convention
  max_tokens?: number
}

function toCompletion(resp: GavioResponse): ChatCompletion {
  return {
    id: resp.traceId,
    object: 'chat.completion',
    model: resp.modelVersion || resp.model,
    choices: [
      { index: 0, message: { role: 'assistant', content: resp.content }, finish_reason: 'stop' },
    ],
    usage: {
      prompt_tokens: resp.usage.promptTokens,
      completion_tokens: resp.usage.completionTokens,
      total_tokens: resp.usage.totalTokens,
    },
    gavio: {
      costUsd: resp.costUsd,
      cacheHit: resp.cacheHit,
      interceptorsFired: resp.interceptorsFired,
    },
  }
}

class Completions {
  constructor(private readonly gw: Gateway) {}

  async create(params: CreateParams): Promise<ChatCompletion> {
    const resp = await this.gw.complete({
      messages: params.messages,
      model: params.model,
      options: { temperature: params.temperature ?? 0.7, maxTokens: params.max_tokens ?? 1024 },
    })
    return toCompletion(resp)
  }
}

/** OpenAI-client-shaped facade over a Gavio Gateway. */
export class GavioOpenAI {
  readonly chat: { completions: Completions }

  constructor(gateway: Gateway) {
    this.chat = { completions: new Completions(gateway) }
  }
}

File: packages/gavio-py/gavio/shim/openai.py
"""OpenAI drop-in shim (F-DX-04) — point existing OpenAI SDK code at Gavio.

    from gavio import Gateway
    from gavio.shim.openai import GavioOpenAI

    client = GavioOpenAI(Gateway.builder().provider("openai").model("gpt-4o").build())
    resp = client.chat.completions.create(
        model="gpt-4o", messages=[{"role": "user", "content": "hi"}]
    )
    print(resp.choices[0].message.content)

The response object mirrors the shape of the OpenAI SDK's ChatCompletion, so
most existing call sites work unchanged — now with PII guard, audit, caching,
etc. running underneath.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from ..gateway import Gateway
from ..response import GavioResponse


@dataclass
class _Message:
    role: str
    content: str


@dataclass
class _Choice:
    index: int
    message: _Message
    finish_reason: str = "stop"


@dataclass
class _Usage:
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int


@dataclass
class ChatCompletion:
    id: str
    model: str
    choices: list[_Choice]
    usage: _Usage
    object: str = "chat.completion"
    gavio: dict[str, Any] = field(default_factory=dict)


def _to_completion(resp: GavioResponse) -> ChatCompletion:
    return ChatCompletion(
        id=resp.trace_id,
        model=resp.model_version or resp.model,
        choices=[_Choice(0, _Message("assistant", resp.content))],
        usage=_Usage(
            prompt_tokens=resp.usage.prompt_tokens,
            completion_tokens=resp.usage.completion_tokens,
            total_tokens=resp.usage.total_tokens,
        ),
        gavio={
            "cost_usd": resp.cost_usd,
            "cache_hit": resp.cache_hit,
            "interceptors_fired": resp.interceptors_fired,
        },
    )


class _Completions:
    def __init__(self, gateway: Gateway) -> None:
        self._gw = gateway

    def create(
        self,
        messages: list[dict],
        model: str | None = None,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> ChatCompletion:
        resp = self._gw.complete_sync(
            messages=messages,
            model=model,
            temperature=temperature,
            max_tokens=max_tokens,
        )
        return _to_completion(resp)

    async def acreate(
        self,
        messages: list[dict],
        model: str | None = None,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> ChatCompletion:
        resp = await self._gw.complete(
            messages=messages,
            model=model,
            temperature=temperature,
            max_tokens=max_tokens,
        )
        return _to_completion(resp)


class _Chat:
    def __init__(self, gateway: Gateway) -> None:
        self.completions = _Completions(gateway)


class GavioOpenAI:
    """OpenAI-client-shaped facade over a Gavio :class:`Gateway`."""

    def __init__(self, gateway: Gateway) -> None:
        self.chat = _Chat(gateway)
